### ISSUE TITLE
Feature Request: Iterator for 'RandomStream'

### ISSUE THREAD

**Issue by LouisJenkinsCS** [chapel-lang/chapel#10708]
One thing that I find myself wishing for is a way to iterate over a `RandomStream`. Currently during my PNNL project, I have been working more and more with the `Random` module, and in fact one of the issues that have set back the project (besides other issues that include but are not limited to issues in #10176) is how tricky it is to iterate in parallel, across multiple locales and tasks. 

The naive way will result in degrading performance, but it can be a trap that newcomers eagerly fall into. The `parSafe` option just introduces a [naive coarse-grained lock](https://github.com/chapel-lang/chapel/blob/master/modules/standard/Random.chpl#L620-L621), and I believe that it is easy for users to think that it will create something scalable. As well if the `RandomStream` is accessed from multiple locales it will be not only subject to additional communication that is non-trivial since [`RandomStream` is a class](https://github.com/chapel-lang/chapel/blob/master/modules/standard/Random.chpl#L533) allocated on a single locale's heap, but also it slows down throughput to the time it takes for a remote task to acquire and release that lock (with round-trip PUT/GET each time). What is worse is that sometimes the user is unable to do anything but use this approach if you need to use a `RandomStream` from a parallel iterator. 

```chpl
var rng = new RandomStream(int, parSafe=true);
forall _ in someArray do rng.getNext();
```

The less naive way is to create per-task random number generator in the cases where you can create one on each task. This is the way I have been doing for a while, being blinded by the promise of performance. The issue with this approach is that it requires the manual creation of tasks, which is impossible to do if you are using a `standalone` or `leader`/`follower` iterator via a `forall`, and even then it can result in really-difficult-to-debug bugs (like, ~2 months for a certain hypergraph generator) due to having *slightly* more duplicates than expected (we're talking like 0.0001% more duplicates than expected), which is enough to skew more extensive calculations. I believe that this issue comes from the fact that all tasks end up using a different seed, but if they naively share the same seed it will result in an increase in the number of duplicates.

```chpl
coforall tid in 1..here.maxTaskPar {
   var rng = new RandomSteam(int);
   for 1..perTaskIterations do rng.getNext();
}
```

One thing that I would like to see from the `Random` module is a way to iterate over the random stream in a parallel and distributed context. One nice extension would be creating a leader-follower iterator where the leader can create a seed that all follower tasks share, where followers `skipToNth` to the offset provided by the leader.

```chpl
// If `randStream` is not specified, it will create a temporary one with the default seed.
iter randomStream(numSamples : int, targetLocales = Locales, randStream : RandomStream = nil);
for rng in randomStream(10) do writeln(rng); // One task, one locale
forall rng in randomStream(10) do writeln(rng); // Parallel & Distributed
forall rng in randomStream(10, targetLocales = [here]) do writeln(rng); // Parallel, one locale
```

As well, it would be nice if it was possible to `zip` the `randomStream` with a distributed array...

```chpl
forall (_, rng) in zip(someArray, randomStream(someArray.size)) do ;
```

**Comment by bradcray** [comment]
> One thing that I find myself wishing for is a way to iterate over a RandomStream.

This is a longstanding request, but one that runs into some limitations in the current parallel iterator design that have yet to be addressed.  There is an undocumented iterator for the PCG random stream named `PCGRandomPrivate_iterate` that can be used for a single parallel iteration, but IIRC, each use of it after that will start over again at the same initial position in the random stream.  I.e., if I consume `n` elements out of it in parallel, the next element I get from it won't be the `n`+1-st, it'll be the initial one again.  You can work around this by manually fast-forwarding the parallel stream between random loops over it, of course... but this is why the interface hasn't been made public (it's not at all intuitive nor what we want).

I didn't have time to read your whole response, but wanted to get you this information quickly in case it helps you before I'm able to go back and spend more time with this issue.

**Comment by mppf** [comment]
I feel like I'm missing something.

There is a *public* and *documented* way to iterate over a random stream even with a forall loop:

https://chapel-lang.org/docs/modules/standard/Random/PCGRandom.html#PCGRandom.RandomStream.iterate

Why is this not sufficient?

**Comment by LouisJenkinsCS** [comment]
Oh? I didn't even see that.

**Comment by LouisJenkinsCS** [comment]
One thing though... is it possible to have the documentation show `pragma "fn returns iterator"` or whatever it is, to change it to show `iter` instead of `proc`

**Comment by bradcray** [comment]
I went looking for this but failed to find it for reasons similar to Louis: I was searching the code for `iter` (and also carrying along some baggage/memories from when I tried and failed to write the iterator case).

@LouisJenkinsCS : I think we'd ultimately want this to be an iterator, but that it isn't because of the limitation I alluded to above.  I'd forgotten that Michael came up with this clever workaround.  I think the desire to show it returning an iterator makes sense and aligns with the longstanding thought about how to make `_iteratorRecord` more first-class / user-facing as you touch on in issue #10361.
